Fix plotting tests for matplotlib v.2.1 and other bugs

ericpre opened the pull request:
### Description of the change
Matplotlib 2.1 introduces some changes, which breaks our plotting tests but at the same time helps/force us to find a few bugs and to fix it!
- we had a recursive loop when drawing plot, [this matplotlib change](https://matplotlib.org/api/api_changes.html#matplotlib-cbook-callbackregistry-process-suppresses-exceptions-by-default) now raise an error -now fixed by disconnecting and re-connecting the `draw_event` when drawing. I was expecting that it will fix the bug mentioned in [#1698](https://github.com/hyperspy/hyperspy/pull/1698#issuecomment-325693283) and #726, but this is not the case...
- change the matplotlib installation in appevyor: use freetype=2.6 from conde-forge instead of the matplotlib wheel from pypi.

### Progress of the PR
- [x] fix recursive loop with `draw_event`,
- [x] avoid create new axis, when not necessary -see [this API change in matplotlib 2.1](https://matplotlib.org/api/api_changes.html#axes-collision-in-figure-add-axes),
- [x] update two baseline images.
- [x] update dev guide regarding the required freetype version for plotting tests,
- [x] ready for review.

ericpre committed:
Create axis in figure only when necessary following matplotlib 2.1 deprecationwarning.

ericpre committed:
Disconnect and re-connect matplotlib `draw_event` to avoid recursive loop occurring in matplotlib 2.1.

ericpre committed:
Remove unused `plot_ticks` attribute in `ImagePlot`.

ericpre committed:
Update baseline image for matplotlib 2.1 and fix issue with qt backend.

ericpre committed:
Set freetype=2.5.5 in appveyor.

ericpre committed:
The matplotlib wheels from pip (on windows) now come with freetype 2.8, which makes the plotting tests fail: switch back to matplotlib from conda with --no-deps option to force to use freetype 2.5.

ericpre committed:
Try freetype=2.6 from conda-forge.

ericpre committed:
Simplify appveyor.yml and update doc.

magnunor commented on hyperspy/drawing/figure.py:
```diff
@@ -51,12 +76,18 @@ def _draw_animated(self):
+                    # fix. Properly fixing the issue involves avoiding
+                    # calling this method too early in the code.
+                    _logger.warning('Please report this warning to the '
+                                    'developpers: `_draw_animated` was called '
```
Typo: developpers -> developers

magnunor commented:
Checking to see if this solves my "recursive loop" plotting issue.

magnunor commented:
This fixes the bug! I suggest we get this merged asap, and release a new patch version (1.3.1). Since I'm guessing anyone currently installing HyperSpy via either PyPI or Anaconda will experience the same bug?

ericpre commented:
It sounds to me that you have just committed yourself to review this one! ;)
This issue occurs only with matplotlib 2.1 and two workarounds are to downgrade matplotlib to 2.0 or to use another backend, so it is not really like HyperSpy is broken.

magnunor commented:
Seems like the failing travis test is due to lxml missing from the Anaconda repos, which should be fixed in RELEASE_next_minor (so I guess we can ignore that failing test).

I'm not sure if I feel qualified enough to comment on the changes in the plotting part, since I'm not that familiar with it... However, it seems to work.

The workarounds are fine, but the bug itself is fairly non-intuitive for non-experienced users. So the majority will probably just give up if they encounter it.

vidartf commented on hyperspy/drawing/figure.py:
```diff
@@ -51,12 +76,18 @@ def _draw_animated(self):
             if canvas.supports_blit:
-                canvas.blit(self.figure.bbox)
+                if hasattr(canvas, 'update'):
+                    canvas.update()
```
I cannot find any official documentation for this function. Could you link it? When will it be there, and what does it do differently than the blit?

to266 replied:
according to [this](https://github.com/matplotlib/matplotlib/pull/9505), is the `update` or `blit` even necessary? 

The relevant bit:
```
 +    In most backends callbacks subscribed to this callback will be
 +    fired after the rendering is complete but before the screen is
 +    updated.  Any extra artists drawn to the canvas's renderer will
 +    be reflected without an explicit call to ``blit``.
 +
 +    .. warning ::
 +
 +       Calling ``canvas.draw`` and ``canvas.blit`` in these callbacks may
 +       not be safe with all backends and may cause infinite recursion.
```

ericpre replied:
This is the `update` method from QtWidget. For some reason, the `blit(self.figure.bbox)` is not working nicely anymore with the qt4 backend (I haven't tested qt5): it does not draw completely the plot but it does it after a draw event is triggered... the `update` method does the job as before though!

francisco-dlp replied:
As this has worked properly until recently, I that that you may have trying to workaround an external bug here. What is triggering the bug, the version of qt4 or matplotlib? Maybe the issue has been fixed externally in the mean time? In any case, would it be possible to simply skip the problematic version in ``setup.py``? If not,  I would suggest to  change the ``if hasattr(canvas, 'update')`` for checking if the backend is qt4 in order not to trigger the ``update`` unnecessarily for other backends.

ericpre replied:
Yes, this was more or less a workaround and not a very good one, quite a few changes have been made for the qt backends in the last few matplotlib releases -not sure this is completely finished! I have not followed what's going on but there are some relevant information in matplotlib/matplotlib#9505 and matplotlib/matplotlib#9199.

ericpre committed:
Fix typo, revert back update/blit workaround.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into fix_plotting_tests_matplotlib_v.2.1

ericpre committed:
Skip matplotlib version 2.1.0 and 2.1.1 to avoid issue with the qt backend.

ericpre committed:
Flush events to plot images, since the last matplotlib version seems to need it...

ericpre committed:
Use freetype=2.6 in travis to make the plotting tests pass.

francisco-dlp committed:
Add comments

The comments were taken directly from https://matplotlib.org/examples/pylab_examples/system_monitor.html

francisco-dlp commented:
I have just added some comments to the line were you flush the events. Other than that it LGTM so...
